Add documentation on the OTel Spring Boot starter

jeanbisutti opened the pull request.

jeanbisutti commented:
This PR could be merged after an OTel Java instrumentation release contains https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/9625

trask commented:
> This PR could be merged after an OTel Java instrumentation release contains [open-telemetry/opentelemetry-java-instrumentation#9625](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/9625)

since it will be ~ a month before next OTel Java Instrumentation release, what do you think of submitting a PR without the DataSource docs first and we can merge that, and a follow-up PR adding in the DataSource docs?

jeanbisutti commented:
> > This PR could be merged after an OTel Java instrumentation release contains [open-telemetry/opentelemetry-java-instrumentation#9625](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/9625)
> 
> since it will be ~ a month before next OTel Java Instrumentation release, what do you think of submitting a PR without the DataSource docs first and we can merge that, and a follow-up PR adding in the DataSource docs?

Documentation on JDBC instrumentation removed

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,53 @@
+  - /docs/instrumentation/java/springboot
+weight: 30
+description: Spring instrumentation for OpenTelemetry Java
+# prettier-ignore
```
```suggestion
cSpell:ignore: springboot autoconfigure 
```

this will make sure that cspell stops complaining

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,53 @@
+# prettier-ignore
+---
+
+You can instrument a Spring Boot application with the [OpenTelemetry Java agent](automatic/_index).
+
+Another option is to use the OpenTelemetry Spring Boot starter. The rest of this page document it.
```
```suggestion
You can automatically instrument a [Spring Boot](https://spring.io/projects/spring-boot) application using the [OpenTelemetry Java agent](automatic/_index), or you can use the OpenTelemetry Spring Boot starter, as outlined below.
```

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,53 @@
+
+## Getting Started
+
+OpenTelemetry Spring Starter is a starter package that allow you to setup distributed tracing.
+
+Check out [opentelemetry-spring-boot-autoconfigure](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-boot-autoconfigure/README.md#features) for the full list of supported libraries and features.
```
As someone who is not involved here, this is confusing, because the link you share brings me to something called "OpenTelemetry Spring Auto-Configuration", which then links to "OpenTelemetry Spring Starter".

Also, people might not know what a "Starter" and what a "Auto-Configuration" is. Do we have anything we can reference here first to clarify what they are going to get here (especially also compared to the agent)

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,53 @@
+
+Add the dependency given below in your project to enable the OpenTelemetry Spring Starter.
+
+You have to replace`OPENTELEMETRY_VERSION` with the latest stable [release](https://search.maven.org/search?q=g:io.opentelemetry).
```
```suggestion
You have to replace`OPENTELEMETRY_VERSION` with the latest stable [release](https://search.maven.org/search?q=g:io.opentelemetry).
```

If the starter is in sync with other releases we should be able to pull in the latest version like we do it on the other pages.

jeanbisutti replied:
It is fixes with 125a3823e30186a39d5bca9ec0437f928591b1cc

svrnm reviewed (changes requested):
Thanks for taking the time to write this down @jeanbisutti! I provided some change requests below.

My major point is that people not familiar with OpenTelemetry might get confused what the differences between this approach and the agent are. This document should either link to another page clarifying that or doing it directly.

jeanbisutti commented:
Hi @svrnm!

Many thanks for your review!

I have worked on addressing your feedback. Could you please review the last changes?

Could you please provide some guidance to fix the CI issues?

Thank you!

svrnm commented:
> I have worked on addressing your feedback. Could you please review the last changes?

I was out of office last week, so apologies for the delay, let me take another pass
 
> Could you please provide some guidance to fix the CI issues?

do not worry too much about them, some of them can be fixed when the PR is ready to be merged.

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,52 @@
+---
+title: Spring Boot
+linkTitle: Spring Boot
+aliases:
+  - /docs/java/getting_started
+  - /docs/java/springboot
+  - /docs/instrumentation/java/springboot
```
```suggestion
```

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
I am not 100% sure yet where in the docs this should live, @open-telemetry/java-approvers, is this "automatic instrumentation" (and could live under Automatic) or is this something different and we need to find a new home for it?

trask replied:
the spring boot starter is a couple of things:
* autoconfiguration of the OpenTelemetry SDK
* autoinstrumentation of a very limited set of libraries (logging and jdbc for now)

users then add [library instrumentations](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md#libraries--frameworks) for other instrumentations that they need (hopefully some more of these can be autoinstrumented in our spring boot starter in the future)

looking at other lanaguages, I think it could make sense under Automatic Instrumentation (you pull it in as a dependency, and then the OTel SDK is automatically set up and a few libraries are automatically instrumented).

this would probably mean splitting Automatic instrumentation in two though:

* Java agent
* Spring Boot

potentially we could also add Quarkus here (a popular spring boot alternative) which ships with some automatic instrumentation similar to what we provide for Spring Boot

svrnm replied:
Thanks for clarification @trask! Let's keep it as it is right now, but consider how we want to have this look like in the future

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,52 @@
+  - /docs/instrumentation/java/springboot
+weight: 30
+description: Spring instrumentation for OpenTelemetry Java
+javaInstrumentationVersion: 1.31.0
```
I think you can inherit this from the _index, @chalin can you provide some guidance here?

chalin replied:
Yes, let me refactor a few things first before we bring in this PR.

chalin replied:
@svrnm see:

- #3503

PR #3503 should be merged first, and then this PR rebased and adapted to using the new key name.

svrnm replied:
PR is merged, so what changes are needed now?

chalin replied:
See https://github.com/open-telemetry/opentelemetry.io/pull/3377#pullrequestreview-1727954944.

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,52 @@
+cSpell:ignore: springboot autoconfigure 
+---
+
+You can automatically instrument a [Spring Boot](https://spring.io/projects/spring-boot) application using the [OpenTelemetry Java agent](automatic/_index) that uses byte code instrumentation.
```
```suggestion
You can automatically instrument a [Spring Boot](https://spring.io/projects/spring-boot) application using the [OpenTelemetry Java agent](../automatic/) that uses byte code instrumentation.
```

svrnm commented on content/en/docs/instrumentation/java/springboot.md:
```diff
@@ -0,0 +1,52 @@
+
+You can automatically instrument a [Spring Boot](https://spring.io/projects/spring-boot) application using the [OpenTelemetry Java agent](automatic/_index) that uses byte code instrumentation.
+
+The OpenTelemetry starter is another way to instrument your Spring Boot application. It is a [Spring Boot starter](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using.build-systems.starters).
```
```suggestion
The OpenTelemetry Spring Boot Starter is another way to instrument your Spring Boot application. It is a [Spring Boot starter](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using.build-systems.starters).
```

I am not 100% sure what the exact name of this should be, but Starter should be capitalized and it should be part of the name that this is for Spring

svrnm reviewed (changes requested):
a few more comments.

chalin commented on content/en/docs/instrumentation/java/springboot.md:
At a minimum the file should be renamed to `spring-boot.md` to match the usual hyphenated naming used elsewhere.

svrnm replied:
@jeanbisutti please rename the file as suggested

svrnm replied:
yes, this will be there automatically

svrnm replied:
you have to move it into `automatic/`

jeanbisutti replied:
Done. Thank you.

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
```diff
@@ -0,0 +1,48 @@
+linkTitle: Spring Boot
+weight: 30
+description: Spring instrumentation for OpenTelemetry Java
+javaInstrumentationVersion: 1.31.0
```
```suggestion
```

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
```diff
@@ -0,0 +1,48 @@
+  <dependency>
+    <groupId>io.opentelemetry.instrumentation</groupId>
+    <artifactId>opentelemetry-spring-boot-starter</artifactId>
+    <version>{{% param javaInstrumentationVersion %}}</version>
```
```suggestion
    <version>{{% param vers.instrumentation %}}</version>
```

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
````diff
@@ -0,0 +1,48 @@
+Add the following dependencies to your gradle.build file:
+
+```groovy
+implementation("io.opentelemetry.instrumentation:opentelemetry-spring-boot-starter:{{% param javaInstrumentationVersion %}}")
````
```suggestion
implementation("io.opentelemetry.instrumentation:opentelemetry-spring-boot-starter:{{% param vers.instrumentation %}}")
```

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
````diff
@@ -0,0 +1,48 @@
+
+You can look at [this](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-boot-autoconfigure/README.md#features) to know the supported libraries and features of the OpenTelemetry auto-configurations.
+
+### Maven
+
+Add the following dependencies to your `pom.xml` file:
+
+```xml
+<dependencies>
+  <dependency>
+    <groupId>io.opentelemetry.instrumentation</groupId>
+    <artifactId>opentelemetry-spring-boot-starter</artifactId>
+    <version>{{% param javaInstrumentationVersion %}}</version>
+  </dependency>
+</dependencies>
+```
+
+### Gradle
+
+Add the following dependencies to your gradle.build file:
+
+```groovy
+implementation("io.opentelemetry.instrumentation:opentelemetry-spring-boot-starter:{{% param javaInstrumentationVersion %}}")
+```
````
Use tabs for the alternatives "Maven" and "Gradle"

jeanbisutti replied:
@chalin I am not sure to follow. In my last commit I have aligned  this Gradle example with another one found on this repository.

jeanbisutti replied:
It should be fine my new commit

chalin reviewed (commented):
See inline suggestions that replace `javaInstrumentationVersion` with `vers.instrumentation`.

jeanbisutti commented:
@chalin @svrnm Is there anything else to do on my side?

chalin commented:
/fix:format

chalin commented:
> @chalin @svrnm Is there anything else to do on my side?

Looking now.

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
```diff
@@ -0,0 +1,59 @@
+cSpell:ignore: springboot autoconfigure
+---
+
+You can automatically instrument a
+[Spring Boot](https://spring.io/projects/spring-boot) application using the
+[OpenTelemetry Java agent](..) that uses byte code instrumentation.
+
+The OpenTelemetry Spring Boot Starter is another way to instrument your Spring
+Boot application. It is a
+[Spring Boot starter](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using.build-systems.starters).
+
+The OpenTelemetry starter is compatible with Spring Boot 2.0, Spring Boot 3.0
+and Spring native.
```
```suggestion
You can use the [OpenTelemetry Java agent](..) with byte code instrumentation to
automatically instrument a [Spring Boot](https://spring.io/projects/spring-boot)
application; or you can also use the OpenTelemetry [Spring Boot starter] to
instrument your application.

[Spring Boot starter]:
  https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using.build-systems.starters

The OpenTelemetry starter is compatible with Spring Boot 2.0 and 3.0, and Spring
native.
```

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
```diff
@@ -0,0 +1,59 @@
+The OpenTelemetry starter is compatible with Spring Boot 2.0, Spring Boot 3.0
+and Spring native.
+
+## Getting Started
```
```suggestion
## Configuration
```

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
```diff
@@ -0,0 +1,59 @@
+
+Add the dependency given below to enable the OpenTelemetry starter.
+
+The OpenTelemetry starter uses OpenTelemetry
+[Spring Boot auto-configurations](https://docs.spring.io/spring-boot/docs/current/reference/html/using.html#using.auto-configuration).
+
+You can look at
+[this](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-boot-autoconfigure/README.md#features)
+to know the supported libraries and features of the OpenTelemetry
+auto-configurations.
```
```suggestion
The OpenTelemetry starter uses OpenTelemetry Spring Boot [auto-configuration].
For details concerning supported libraries and features of the OpenTelemetry
auto-configuration, see the configuration [README].

[auto-configuration]:
  https://docs.spring.io/spring-boot/docs/current/reference/html/using.html#using.auto-configuration
[README]:
  https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-boot-autoconfigure/README.md#features
```

chalin commented on content/en/docs/instrumentation/java/automatic/spring-boot.md:
````diff
@@ -0,0 +1,59 @@
+to know the supported libraries and features of the OpenTelemetry
+auto-configurations.
+
+### Maven
+
+Add the following dependencies to your `pom.xml` file:
+
+```xml
+<dependencies>
+	<dependency>
+		<groupId>io.opentelemetry.instrumentation</groupId>
+		<artifactId>opentelemetry-spring-boot-starter</artifactId>
+		<version>{{% param vers.instrumentation %}}</version>
+	</dependency>
+</dependencies>
+```
+
+### Gradle
+
+Add the following dependencies to your gradle.build file:
+
+```groovy
+dependencies {
+	implementation('io.opentelemetry.instrumentation:opentelemetry-spring-boot-starter:{{% param vers.instrumentation %}}')
+}
+```
````
~~~suggestion
{{< tabpane text=true >}} {{% tab header="Maven (`pom.xml`)" lang=Maven %}}

```xml
<dependencies>
	<dependency>
		<groupId>io.opentelemetry.instrumentation</groupId>
		<artifactId>opentelemetry-spring-boot-starter</artifactId>
		<version>{{% param vers.instrumentation %}}</version>
	</dependency>
</dependencies>
```

{{% /tab %}} {{% tab header="Gradle (`gradle.build`)" lang=Gradle %}}

```groovy
dependencies {
	implementation('io.opentelemetry.instrumentation:opentelemetry-spring-boot-starter:{{% param vers.instrumentation %}}')
}
```

{{% /tab %}} {{< /tabpane>}}
~~~

chalin reviewed (commented):
A (hopefully final) copyedit pass, adding tabbed panes too.

jeanbisutti commented:
Thank you @svrnm and @chalin for the review and the help.
